src/objective-c/GRPCClient/private/GRPCWrappedCall.m

```diff
@@ -0,0 +1,341 @@
+#import "NSDictionary+GRPC.h"
+#import "NSData+GRPC.h"
+#import "NSError+GRPC.h"
+
+@implementation GRPCOpSendMetadata{
+  void(^_handler)(void);
+  grpc_metadata *_send_metadata;
+  size_t _count;
+}
+
+- (instancetype)init {
+  return [self initWithMetadata:nil handler:nil];
+}
+
+- (instancetype)initWithMetadata:(NSDictionary *)metadata handler:(void (^)(void))handler {
+  if (self = [super init]) {
+    if (metadata) {
+      [metadata grpc_getMetadataArray:&_send_metadata];
+      _count = metadata.count;
+    } else {
+      _send_metadata = NULL;
+      _count = 0;
+    }
+    _handler = handler;
+  }
+  return self;
+}
+
+- (void)getOp:(grpc_op *)op {
+  op->op = GRPC_OP_SEND_INITIAL_METADATA;
+  op->data.send_initial_metadata.count = _count;
+}
+
+- (void (^)(void))opProcessor {
+  return ^{
+    gpr_free(_send_metadata);
+    if (_handler) {
+      _handler();
+    }
+  };
+}
+
+@end
+
+@implementation GRPCOpSendMessage{
+  void(^_handler)(void);
+  grpc_byte_buffer *_byte_buffer;
+}
+
+- (instancetype)init {
+  return [self initWithMessage:nil handler:nil];
+}
+
+- (instancetype)initWithMessage:(NSData *)message handler:(void (^)(void))handler {
+  if (!message) {
+    [NSException raise:NSInvalidArgumentException format:@"message cannot be null"];
+  }
+  if (self = [super init]) {
+    _byte_buffer = [message grpc_byteBuffer];
+    _handler = handler;
+  }
+  return self;
+}
+
+- (void)getOp:(grpc_op *)op {
+  op->op = GRPC_OP_SEND_MESSAGE;
+  op->data.send_message = _byte_buffer;
+}
+
+- (void (^)(void))opProcessor {
+  return ^{
+    gpr_free(_byte_buffer);
+    if (_handler) {
+      _handler();
+    }
+  };
+}
+
+@end
+
+@implementation GRPCOpSendClose{
+  void(^_handler)(void);
+}
+
+- (instancetype)init {
+  return [self initWithHandler:nil];
+}
+
+- (instancetype)initWithHandler:(void (^)(void))handler {
+  if (self = [super init]) {
+    _handler = handler;
+  }
+  return self;
+}
+
+- (void)getOp:(grpc_op *)op {
+  op->op = GRPC_OP_SEND_CLOSE_FROM_CLIENT;
+}
+
+- (void (^)(void))opProcessor {
+  return ^{
+    if (_handler) {
+      _handler();
+    }
+  };
+}
+
+@end
+
+@implementation GRPCOpRecvMetadata{
+  void(^_handler)(NSDictionary *);
+  grpc_metadata_array *_recv_initial_metadata;
+}
+
+- (instancetype) init {
+  return [self initWithHandler:nil];
+}
+
+- (instancetype) initWithHandler:(void (^)(NSDictionary *))handler {
+  if (self = [super init]) {
+    _handler = handler;
+    _recv_initial_metadata = gpr_malloc(sizeof(grpc_metadata_array));
+    grpc_metadata_array_init(_recv_initial_metadata);
+  }
+  return self;
+}
+
+- (void)getOp:(grpc_op *)op {
+  op->op = GRPC_OP_RECV_INITIAL_METADATA;
+  op->data.recv_initial_metadata = _recv_initial_metadata;
+}
+
+- (void (^)(void))opProcessor {
+  return ^{
+    NSDictionary *metadata = [NSDictionary
+                              grpc_dictionaryFromMetadata:_recv_initial_metadata->metadata
+                              count:_recv_initial_metadata->count];
```

Comment:
(Indentation looks odd here, but I'm not sure what's the correct one; make sure to pass these lines through XCode's ctrl+i)